### PR TITLE
Fix sklearn autologging x distutils issue

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/12847?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12847/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12847
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Fix for https://github.com/scikit-learn/scikit-learn/issues/26992.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

The following code works fine:

```python
from sklearn.cluster import KMeans
from sklearn.datasets import load_iris

import mlflow

model = KMeans().fit(*(load_iris(return_X_y=True)))

with mlflow.start_run():
    mlflow.sklearn.log_model(model, "model")

mlflow.sklearn.autolog()

```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/sklearn/test_sklearn_autolog.py [hunk 1]
```diff
@@ -1806,3 +1808,24 @@ def test_autolog_emits_warning_message_when_pos_label_used_for_multilabel():
             "Metric error: Target is multiclass but average='binary'. Please choose another "
             "average setting, one of [None, 'micro', 'macro', 'weighted']."
         )
+
+
+def test_sklearn_distutils_issue(monkeypatch, tmp_path):
```
Comment (harupy) [inline tests/sklearn/test_sklearn_autolog.py:1813]: The fix is pending. We must ensure this test really fails.

Comment (harupy) [reply tests/sklearn/test_sklearn_autolog.py]: Interesting. I could not repro the issue on CI.

Comment (harupy) [reply tests/sklearn/test_sklearn_autolog.py]: Failed! 
https://github.com/mlflow/mlflow/actions/runs/10192119881/job/28194484685?pr=12847

Comment (harupy) [reply tests/sklearn/test_sklearn_autolog.py]: It seems it's fixed on sklearn master branch.

File: mlflow/utils/environment.py [hunk 2]
```diff
@@ -265,18 +259,10 @@ def _mlflow_conda_env(
...
-def _get_pip_version():
-    """
-    Returns:
-        The version of ``pip`` that is installed in the current environment,
-        or ``None`` if ``pip`` is not currently installed / does not have a
-        ``__version__`` attribute.
-    """
+def _get_package_version(package_name: str) -> Optional[str]:
     try:
-        import pip
```
Comment (harupy) [inline mlflow/utils/environment.py:276]: turns out importing pip can mess up setuptools.

Comment (WeichenXu123) [reply mlflow/utils/environment.py:276]: nice finding!

**Comment by WeichenXu123** [comment]
Some test failures:
```
package_name = 'wheel'

    def _get_package_version(package_name: str) -> Optional[str]:
        try:
            return importlib.metadata.version(package_name)
>       except importlib.metadata.version.PackageNotFoundError:
E       AttributeError: 'function' object has no attribute 'PackageNotFoundError'
```

**Comment by harupy** [comment]
@WeichenXu123 Fixed that error

**Review by WeichenXu123 — APPROVED** [review]
LGTM

**Pull request merged**